### PR TITLE
Changes necessary for hot reload to work

### PR DESCRIPTION
**Pull request by dsnopek**
These are the changes necessary to work with Godot PR https://github.com/godotengine/godot/pull/80284

### CODE REVIEW

**Comment by dsnopek** [comment]
Hm, this worked in my testing because my test classes declared their own constructor without any arguments. This is failing the tests because they don't, and just rely on the default constructor being inherited from the parent class. But now that this PR adds a secondary constructor, the default won't be inherited anymore. :-/

I'm not yet sure what a good solution would be...

**Comment by dsnopek** [comment]
I came up with a "hack" to workaround the problem described in my last comment.

Rather than adding a new constructor that can handle recreating instances, it adds the info about any instances being recreated to a global list, and then the usual constructor will check if this instance is meant to be recreated, and if so, do that process instead. Since this is a little ugly, I've wrapped this in a `HOT_RELOAD_ENABLED` define, that by default will only be set for debug and editor builds. The thing I'm most unsure about is the data structure used to hold the list, so suggestions on better ways to do that welcome.

**Comment by dsnopek** [comment]
Taking this one out of draft because the Godot PR has also come out of draft, but it won't pass tests (and can't be merged) until Godot PR https://github.com/godotengine/godot/pull/80284 is merged

**Comment by dsnopek** [comment]
Temporarily rebasing on PR https://github.com/godotengine/godot-cpp/pull/1239 to keep this compatible with the Godot PR. Hopefully, that one will be merged soon-ish and this can go back to being a single commit

**Comment by DmitriySalnikov** [comment]
I have updated the library. Now there is no crash in `godot::Wrapped::Wrapped`, but there remained random crashes after several reloads.
I also noticed that only `_enter_tree` is called for `EditorPlugin`, and `_exit_tree`, `_disable_plugin` and `_enable_plugin` don't react to reloading in any way (but I can use a destructor to deinitialize). Perhaps it is implied that it is a `Node` and should simply replace the pointer to the new library. But if so, then somehow I need to know that it doesn't need to be deleted during deinitialization.

Also a re-registered singleton still causes:
(here I added the `print` of the singleton id from the library and script)
```
GD: <Freed Object>
SCRIPT ERROR: Invalid call. Nonexistent function 'clear_graphs' in base 'previously freed'.
          at: _process (res://examples_dd3d/DebugDrawDemoScene.gd:73)
[Wrapped:0]
```

**Comment by DmitriySalnikov** [comment]
I want to add information about singleton.
If I replace the standard singleton getting via the class name with `Engine.get_singleton`, then the calls work without errors.
```gdscript
# work
Engine.get_singleton(&"DebugDraw3D").draw_box(box_pos, Vector3(1, 2, 1), Color(0, 1, 0))
# doesn't work
DebugDraw3D.draw_box(box_pos, Vector3(1, 2, 1), Color(0, 1, 0))
```

https://github.com/godotengine/godot-cpp/assets/7782218/88c99b60-e21d-49f0-acd8-eb3ed3eabc81

Also, if I use my bindings for C#, then after a reload everything also continues to work. Because I check the validity of the cached instance and try to get it if it is `null` or `invalid`.

```cs
    private static GodotObject _instance;
    public static GodotObject Instance
    {
        get
        {
            if (!GodotObject.IsInstanceValid(_instance))
            {
                _instance = Engine.GetSingleton("DebugDraw3D");
            }
            return _instance;
        }
    }
```

It turns out that GDScript just doesn't update the singleton cache somewhere.

**Comment by dsnopek** [comment]
@DmitriySalnikov: Thanks to all your testing and reports I've managed to find and fix a whole bunch of bugs!

> I also noticed that only `_enter_tree` is called for `EditorPlugin`, and `_exit_tree`, `_disable_plugin` and `_enable_plugin` don't react to reloading in any way (but I can use a destructor to deinitialize).

This should be fixed with the latest versions of both PRs: you should get `_exit_tree` when the editor plugin is being deinitialized, before the destructor runs. I don't know that `_disable_plugin` and `_enable_plugin` are appropriate here? The config for the project hasn't actually changed (ie. the plugin would still be enabled if it was enabled before the reload). But we need to look at more editor plugins and see what they're expecting to happen with regard to those virtual functions.

> I want to add information about singleton.
> If I replace the standard singleton getting via the class name with `Engine.get_singleton`, then the calls work without errors.

Hm, interesting, good find! I haven't had a chance to look into this one yet, but I guess we'd need the Godot PR to signal to GDScript to clear its singleton cache somehow.

**Comment by DmitriySalnikov** [comment]
> you should get `_exit_tree` when the editor plugin is being deinitialized

Thanks, I've already checked it out. Now the plugin is removed as if the engine was closed in the usual way without a `reloadable`.

File: gdextension/gdextension_interface.h [hunk 1]
```diff
@@ -267,6 +267,7 @@ typedef void (*GDExtensionClassUnreference)(GDExtensionClassInstancePtr p_instan
 typedef void (*GDExtensionClassCallVirtual)(GDExtensionClassInstancePtr p_instance, const GDExtensionConstTypePtr *p_args, GDExtensionTypePtr r_ret);
 typedef GDExtensionObjectPtr (*GDExtensionClassCreateInstance)(void *p_class_userdata);
 typedef void (*GDExtensionClassFreeInstance)(void *p_class_userdata, GDExtensionClassInstancePtr p_instance);
+typedef GDExtensionClassInstancePtr (*GDExtensionClassRecreateInstance)(void *p_userdata, GDExtensionObjectPtr p_object);
```
Comment (BastiaanOlij) [inline gdextension/gdextension_interface.h:270]: Rename `p_userdata` to `p_class_userdata`

**Review by BastiaanOlij — APPROVED** [review]
LGTM, need to wait on upstream merge and nitpicky rename :)

File: gdextension/gdextension_interface.h (hunk 1 above)
Comment (dsnopek) [reply gdextension/gdextension_interface.h]: Fixed in the latest push!

**Pull request merged**